File: python/src/arker/modal/_process.py
"""`ContainerProcess` — return type of `sandbox.exec()`.

Mirrors modal's `ContainerProcess`: `.wait()`, `.poll()`, `.kill()`,
`.returncode`, and `.stdout` / `.stderr` / `.stdin` stream handles.
Backed by an Arker background run (`run(..., background=True)`).
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING, Iterator

from ..computer import ArkerError
from ._types import SandboxError, translate_arker_error

if TYPE_CHECKING:
    from ._sandbox import Sandbox


POLL_BASE_S = 0.2
POLL_MAX_S = 1.0


def _next_delay(last: float) -> float:
    return min(POLL_MAX_S, max(POLL_BASE_S, last * 1.5))


class StreamReader:
    """Mirrors `modal.io_streams.StreamReader`. The real one is async + line-
    oriented over gRPC; ours is a thin polling wrapper over Arker `run_status`.

    `read()` blocks until the process finishes and returns the full output.
    Iterating yields output lines as they appear (poll-based, not real-time).
    """

    def __init__(self, process: "ContainerProcess", which: str, text: bool = True) -> None:
        self._process = process
        self._which = which  # "stdout" or "stderr"
        self._text = text
        self._cursor = 0

    def read(self) -> str | bytes:
        """Block until the process finishes; return the full stream output."""
        self._process.wait()
        data = self._process._snapshot()[self._which]
        return data if not self._text else data.decode("utf-8", errors="replace")

    def __iter__(self) -> Iterator[str | bytes]:
        """Poll-yield lines as the underlying run produces them."""
        delay = POLL_BASE_S
        buffer = b""
        while True:
            snap = self._process._snapshot()
            new = snap[self._which][self._cursor:]
            self._cursor = len(snap[self._which])
            if new:
                buffer += new
                while b"\n" in buffer:
                    line, _, buffer = buffer.partition(b"\n")
                    line += b"\n"
                    yield line.decode("utf-8", errors="replace") if self._text else line
            if snap["completed"]:
                if buffer:
                    yield buffer.decode("utf-8", errors="replace") if self._text else buffer
                return
            time.sleep(delay)
            delay = _next_delay(delay)


class StreamWriter:
    """Placeholder for `sandbox.stdin`. Arker has no non-PTY stdin primitive,
    so writes raise (loud over silent). Real PTY stdin needs WS — out of scope."""

    def __init__(self, _process: "ContainerProcess") -> None:
        pass

    def write(self, _data: bytes) -> None:
        raise NotImplementedError(
            "arker.modal: ContainerProcess.stdin.write is not supported — "
            "Arker has no non-PTY stdin primitive."
        )

    def drain(self) -> None:
        raise NotImplementedError(
            "arker.modal: ContainerProcess.stdin.drain is not supported."
        )

    def write_eof(self) -> None:
        raise NotImplementedError(
            "arker.modal: ContainerProcess.stdin.write_eof is not supported."
        )


class ContainerProcess:
    def __init__(self, sandbox: "Sandbox", run_id: str, text: bool = True) -> None:
        self._sandbox = sandbox
        self._run_id = run_id
        self._text = text
        self._returncode: int | None = None
        self._final_stdout: bytes = b""
        self._final_stderr: bytes = b""
        self.stdout = StreamReader(self, "stdout", text=text)
        self.stderr = StreamReader(self, "stderr", text=text)
        self.stdin = StreamWriter(self)

    @property
    def returncode(self) -> int | None:
        return self._returncode

    def poll(self) -> int | None:
        """Non-blocking returncode check. Returns None if still running."""
        if self._returncode is not None:
            return self._returncode
        try:
            status = self._sandbox._computer.run_status(self._run_id)
        except ArkerError as error:
            raise translate_arker_error(error) from error
        self._final_stdout = status.stdout
        self._final_stderr = status.stderr
        if status.completed:
            self._returncode = status.exit_code if status.exit_code is not None else -1
        return self._returncode

    def wait(self) -> int:
        """Block until the process finishes; return the returncode."""
        delay = POLL_BASE_S
        while self.poll() is None:
            time.sleep(delay)
            delay = _next_delay(delay)
        assert self._returncode is not None
        return self._returncode

    def kill(self) -> None:
        """Cancel the underlying run; ignore errors since the process may
        already be gone."""
        try:
            self._sandbox._computer.cancel_run(self._run_id)
        except ArkerError:
            pass

    def terminate(self) -> None:
        """Alias for kill — modal has both names on `ContainerProcess`."""
        self.kill()

    def _snapshot(self) -> dict:
        """Internal: re-poll status and return both streams + completion."""
        if self._returncode is not None:
            return {
                "stdout": self._final_stdout,
                "stderr": self._final_stderr,
                "completed": True,
            }
        try:
            status = self._sandbox._computer.run_status(self._run_id)
        except ArkerError as error:
            raise translate_arker_error(error) from error
        if status.completed and self._returncode is None:
            self._returncode = status.exit_code if status.exit_code is not None else -1
            self._final_stdout = status.stdout
            self._final_stderr = status.stderr
        return {
            "stdout": status.stdout,
            "stderr": status.stderr,
            "completed": status.completed,
        }

File: typescript/src/modal/process.ts
import type { RunStatusResponse } from "../index.js";
import type { Sandbox } from "./sandbox.js";
import { InvalidError, translateArkerError } from "./types.js";

const POLL_BASE_MS = 200;
const POLL_MAX_MS = 1_000;

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

function nextDelay(last: number): number {
  return Math.min(POLL_MAX_MS, Math.max(POLL_BASE_MS, last * 1.5));
}

function decode(bytes: Uint8Array): string {
  return new TextDecoder("utf-8", { fatal: false }).decode(bytes);
}

function decodeStreamBytes(value: unknown, encoding: unknown): Uint8Array {
  if (value instanceof Uint8Array) return value;
  if (typeof value !== "string") return new Uint8Array();
  if (encoding === "base64") {
    if (typeof Buffer !== "undefined") return new Uint8Array(Buffer.from(value, "base64"));
    const binary = atob(value);
    const out = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
    return out;
  }
  return new TextEncoder().encode(value);
}

/** Map Arker's exit_code into modal's `128 + signal` encoding. Negative
 * exit codes (Python/Go subprocess convention for signal kills) become the
 * positive form modal returns; null/undefined → -1. */
export function normalizeReturncode(raw: number | null | undefined): number {
  if (raw == null) return -1;
  if (raw < 0) return 128 + Math.abs(raw);
  return raw;
}

function readStream(status: RunStatusResponse, which: "stdout" | "stderr"): Uint8Array {
  const s = status as Record<string, unknown>;
  const value = s[which];
  const encoding = s[`${which}_encoding`];
  return decodeStreamBytes(value, encoding);
}

/** Poll-based stream reader. `read()` blocks until the process finishes;
 * `for await` yields lines as they appear per poll cycle. */
export class StreamReader {
  private readonly proc: ContainerProcess;
  private readonly which: "stdout" | "stderr";
  private readonly text: boolean;
  private cursor = 0;

  constructor(proc: ContainerProcess, which: "stdout" | "stderr", text: boolean) {
    this.proc = proc;
    this.which = which;
    this.text = text;
  }

  async read(): Promise<string | Uint8Array> {
    await this.proc.wait();
    const snap = await this.proc._snapshot();
    const data = snap[this.which];
    return this.text ? decode(data) : data;
  }

  async *[Symbol.asyncIterator](): AsyncIterableIterator<string | Uint8Array> {
    let delay = POLL_BASE_MS;
    let buffer = new Uint8Array();
    while (true) {
      const snap = await this.proc._snapshot();
      const stream = snap[this.which];
      const newSlice = stream.subarray(this.cursor);
      this.cursor = stream.length;
      if (newSlice.length > 0) {
        const merged = new Uint8Array(buffer.length + newSlice.length);
        merged.set(buffer);
        merged.set(newSlice, buffer.length);
        buffer = merged;
        let nl = buffer.indexOf(0x0a);
        while (nl !== -1) {
          const line = buffer.subarray(0, nl + 1);
          buffer = buffer.subarray(nl + 1);
          yield this.text ? decode(line) : line;
          nl = buffer.indexOf(0x0a);
        }
      }
      if (snap.completed) {
        if (buffer.length > 0) yield this.text ? decode(buffer) : buffer;
        return;
      }
      await sleep(delay);
      delay = nextDelay(delay);
    }
  }
}

/** Placeholder for `process.stdin`. Arker has no non-PTY stdin primitive. */
export class StreamWriter {
  async write(_data: Uint8Array): Promise<void> {
    throw new Error("arker.modal: ContainerProcess.stdin.write is not supported — Arker has no non-PTY stdin primitive.");
  }
  async drain(): Promise<void> {
    throw new Error("arker.modal: ContainerProcess.stdin.drain is not supported.");
  }
  async writeEof(): Promise<void> {
    throw new Error("arker.modal: ContainerProcess.stdin.writeEof is not supported.");
  }
}

export class ContainerProcess {
  private readonly sbx: Sandbox;
  private readonly runId: string;
  private readonly text: boolean;
  private _returncode: number | null = null;
  // `Uint8Array` without an explicit ArrayBuffer parameter — Buffer.from()
  // returns Uint8Array<ArrayBufferLike> under newer @types/node strict typing.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private finalStdout: any = new Uint8Array();
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private finalStderr: any = new Uint8Array();

  readonly stdout: StreamReader;
  readonly stderr: StreamReader;
  readonly stdin: StreamWriter;

  constructor(sbx: Sandbox, runId: string, text: boolean) {
    this.sbx = sbx;
    this.runId = runId;
    this.text = text;
    this.stdout = new StreamReader(this, "stdout", text);
    this.stderr = new StreamReader(this, "stderr", text);
    this.stdin = new StreamWriter();
  }

  /** Matches modal: throws InvalidError until `.wait()` resolves.
   * Use `.poll()` for the non-throwing check that returns null while running. */
  get returncode(): number {
    if (this._returncode === null) {
      throw new InvalidError(
        "You must call wait() before accessing the returncode. " +
          "To poll for the status of a running process, use poll() instead.",
      );
    }
    return this._returncode;
  }

  async poll(): Promise<number | null> {
    if (this._returncode !== null) return this._returncode;
    let status: RunStatusResponse;
    try {
      status = await this.sbx._computer.runStatus(this.runId);
    } catch (error) {
      throw translateArkerError(error);
    }
    this.finalStdout = readStream(status, "stdout");
    this.finalStderr = readStream(status, "stderr");
    if ((status as { completed?: boolean }).completed) {
      this._returncode = normalizeReturncode((status as { exit_code?: number | null }).exit_code);
    }
    return this._returncode;
  }

  async wait(): Promise<number> {
    let delay = POLL_BASE_MS;
    while ((await this.poll()) === null) {
      await sleep(delay);
      delay = nextDelay(delay);
    }
    return this._returncode!;
  }

  async kill(): Promise<void> {
    try {
      await this.sbx._computer.cancelRun(this.runId);
    } catch {
      // Ignore: process may already be gone.
    }
  }

  async terminate(): Promise<void> {
    await this.kill();
  }

  /** @internal poll status snapshot for the StreamReader. */
  async _snapshot(): Promise<{ stdout: Uint8Array; stderr: Uint8Array; completed: boolean }> {
    if (this._returncode !== null) {
      return { stdout: this.finalStdout, stderr: this.finalStderr, completed: true };
    }
    let status: RunStatusResponse;
    try {
      status = await this.sbx._computer.runStatus(this.runId);
    } catch (error) {
      throw translateArkerError(error);
    }
    const completed = !!(status as { completed?: boolean }).completed;
    const stdoutBytes = readStream(status, "stdout");
    const stderrBytes = readStream(status, "stderr");
    if (completed && this._returncode === null) {
      this._returncode = normalizeReturncode((status as { exit_code?: number | null }).exit_code);
      this.finalStdout = stdoutBytes;
      this.finalStderr = stderrBytes;
    }
    return { stdout: stdoutBytes, stderr: stderrBytes, completed };
  }
}
